CorsixTH/Src/th_sound.cpp

```diff
@@ -71,11 +71,19 @@ const char* sound_archive::get_sound_name(size_t iIndex) const
     return sound_files[iIndex].sound_name;
 }
 
-#define FOURCC(c1, c2, c3, c4) \
-    ( static_cast<uint32_t>(static_cast<uint8_t>(c1) <<  0) \
-    | static_cast<uint32_t>(static_cast<uint8_t>(c2) <<  8) \
-    | static_cast<uint32_t>(static_cast<uint8_t>(c3) << 16) \
-    | static_cast<uint32_t>(static_cast<uint8_t>(c4) << 24) )
+constexpr uint32_t fourcc(const char c1, const char c2, const char c3, const char c4)
+{
+    return (
+        static_cast<uint32_t>(static_cast<uint8_t>(c1) <<  0)
+        | static_cast<uint32_t>(static_cast<uint8_t>(c2) <<  8)
+        | static_cast<uint32_t>(static_cast<uint8_t>(c3) << 16)
+        | static_cast<uint32_t>(static_cast<uint8_t>(c4) << 24));
```

Comment:
Is this defined, shifting with a larger amount than the width of the data. Perhaps simply cast directly to `uint32_t` ?

Reply:
Good catch!

After a deep dive in the specification, I believe this is implementation defined in C++14 (not undefined) for systems with 32bit integers. In C++11 and earlier it would be undefined, and on C++20 I believe it would be defined as long as the target's int type is > 32bit. All of that assumes that it treats int like one of the fixed width integer types (2s complement is mandated for them, but not for e.g. int, short, etc. directly)

I'll change it.